Wolf UI

ABeltramo opened the pull request:
Plugging together https://github.com/games-on-whales/wolf-ui with Wolf.
WIP will update the PR once more things are ready

ABeltramo committed:
feat: trying to plug together wolf-ui with wolf

ABeltramo committed:
feat: added env WOLF_SESSION_ID

ABeltramo committed:
feat: added profiles and changed config format

Also, changed Moonlight app ID logic to avoid cache.

fixes #140

ABeltramo committed:
fix: implemented APIs for profiles and fixed legacy app endpoints.cpp

fixed tests

ABeltramo committed:
feat: updated OpenAPI spec

ABeltramo commented:
# Introducing profiles

After a lot of discussion in the dev channels, I've added a new concept in Wolf: `profiles`. This will be particularly useful for [wolf-ui](https://github.com/games-on-whales/wolf-ui) which will be the new entrypoint when starting a Moonlight streaming session and will allow to:
 - easily pick up a profile and start an app (this will allow you to carry over savegames and state between different Moonlight clients)
 - easily join another streaming session for co-op playing.

The format in `config.toml` is the following

```toml
[[profiles]]
id = 'moonlight-profile-id'

    [[profiles.apps]]
    title = 'Wolf UI'

        [profiles.apps.runner]
        run_cmd = '/home/ale/repos/wolf-ui/src/wolf-ui'
        type = 'process'

    [[profiles.apps]]
    start_audio_server = false
    start_virtual_compositor = false
    title = 'Test ball'

        [profiles.apps.audio]
        source = 'audiotestsrc wave=ticks is-live=true'

        [profiles.apps.runner]
        run_cmd = "sh -c \"while :; do echo 'running...'; sleep 10; done\""
        type = 'process'

        [profiles.apps.video]
        source = '''videotestsrc pattern=ball flip=true is-live=true !
                           video/x-raw, framerate={fps}/1''' 

[[profiles]]
id = 'user'
name = 'User'
apps = []
```

We'll automatically migrate from the current `config.toml` by:
 - Adding the hardcoded `moonlight-profile-id` which will represent the apps that are going to be shown in the Moonlight UI
 - Adding a default `user` profile and copying over all the previously defined `apps` in here.

## APIs

The currently available `/api/v1/apps` will now control the apps that will be shown in the Moonlight UI. There are 3 new endpoints to control profiles:
 - `GET /api/v1/profiles` returns a list of all defined profiles (most notably it wouldn't return the special "moonlight" profile, you'll have to use `/api/v1/apps` to interact with that one)
 - `POST /api/v1/profiles/add` takes a full profile object
 - `POST /api/v1/profiles/remove` takes just an `id`: the ID of the profile to be deleted

Hitting any of the `apps` or `profiles` endpoints will also automatically save the changes into the `config.toml` (if any).

I've yet to define how to associate `streaming` sessions with `profiles` and how that will play with CO-OP sessions, I'll start working on that now.

@Skerga, @salty2011 please let me know what you guys think and what else you would like to see added!

ABeltramo committed:
feat: coded up full lobbies support!

Removed child_session runner

ABeltramo committed:
fix: upgraded dependencies, cmake warnings and docker build

ABeltramo committed:
feat: implemented lobby APIs

ABeltramo committed:
fix: added default name to interpipesrc

ABeltramo commented:
# Introducing Lobbies

On top of `profiles` we are also introducing Lobbies, here are the highlights:

## Key Features

- **Easy Co-op Session Joining**:
  - Users can now seamlessly join cooperative game sessions from different Moonlight clients

- **Session Continuity Across Devices**:
  - Start a gaming session on one device and continue it on another one without losing progress.

- **Watch Parties**:
  - This will also make possible "watch parties" where users can observe game sessions without interacting. This is perfect for spectators who want to follow along without affecting the gameplay.

ABeltramo committed:
fix: compilation error on gcc, clang-format

ABeltramo committed:
Committing clang-format changes

ABeltramo committed:
fix: implemented stop when everyone leaves, fixed tests

ABeltramo committed:
feat: implemented PIN for lobbies

ABeltramo committed:
feat: exposed lobby_id in CreateLobbyEvent

ABeltramo committed:
fix: convert Moonlight session IDs to String

ABeltramo committed:
fix: use unique interpipesrc naming, properly fix loading old config.toml

ABeltramo committed:
fix: failing tests

ABeltramo committed:
feat: added PIN for profiles and link to Sessions via started_by_profile_id

Skerga committed:
re-added multi_user flag to allow hiding of not normaly joinable singleplayer lobbies

Skerga commented:
Here is my list of things that would still be required for me to deem this branch beta test ready.

- [x] A rebase of the branch.
- [x] Entry in the config defining how an Apps save folder should be named. Possibly something like `directory_name` that 
       can be optional and would default to either the `title` or the `runner.name`.
       this includes maybe defining how wolf-ui should structure profile data. the currently implemented way is: 
       `(/etc/wolf/)profile-data/{profile_id}/{app.runner.name}`
- [x] The same as above but for co-op lobbies, in case a user would like to keep a separate install for co-op play,
       would default to the one above if not used.

Optional stuff that would be nice to have but can wait if necessary:

- [ ] A Keybind that would allow to return back to Wolf-Ui without directly stopping the Lobby the app is running in, 
       including leaving the lobby. (Wolf-UI has logic that allows a user to rejoin a lobby, so switching devices would be
       easier.)
- [x] Setting a Pin for a Profile. currently all profiles are open by default.
- [x] Accessing Images defined in `icon_png_path` via the wolf api if its a local png. this way i could drop access to the folder
       containing the images from wolf-ui's requirements. wolf-ui is capable of using urls by itself and does already access
       the icons on github if `icon_png_path` is not defined.
- [x] API call to pull and update docker images, with some form of progress feedback. This would allow me to keep the
       current update function without the need to access the docker socket and allow wolf-ui to be run less privileged.
- [x] an API endpoint to check if an Apps Docker Image exists on Disk, see reason above.

ABeltramo committed:
fix: rebase fixes

ABeltramo committed:
fix: docker build

ABeltramo committed:
fix: bring back lobbies!

ABeltramo committed:
fix: multi-user lobbies and gstreamer bus warning

ABeltramo committed:
Committing clang-format changes

ABeltramo committed:
CI: setup dependencies to run Gstreamer pipeline in tests

ABeltramo committed:
fix: lobbies logic, sound and edge cases

ABeltramo committed:
Committing clang-format changes

ABeltramo committed:
feat: added /api/v1/utils/get-icon

Unified code with Moonlight HTTP API

ABeltramo committed:
fix: joypad plug into session, early cleanup of closed SSE sockets

ABeltramo committed:
fix: get_icon check response code before returning the content

ABeltramo committed:
CI: use latest stable Rust

ABeltramo committed:
feat: implement progress callback when pulling via Docker APIs

ABeltramo committed:
feat: added thread pool for Wolf API socket

ABeltramo committed:
feat: added Docker API endpoint wrappers in Wolf

ABeltramo committed:
fix: leave SSE connection open

ABeltramo committed:
fix: avoid duplicate client pairing
fixes #211

ABeltramo committed:
fix: thread count in API thread pool

ABeltramo committed:
fix: API move CURL calls in a background thread

ABeltramo committed:
fix: avoid passing string_view to thread

ABeltramo committed:
feat: added Docker pull events in SSE stream

ABeltramo committed:
fix: added Docker pull SSE event, avoid closing connections, updated default for WolfUI

ABeltramo committed:
fix: new proper defaults, correctly setup the API socket, added icon for Lobbies

ABeltramo committed:
fix: derive WOLF_CFG_FOLDER from HOST_APPS_STATE_FOLDER and wrong default steam image

ABeltramo committed:
fix: missing Lobbies icon from the APIs

ABeltramo committed:
fix: expose LeaveLobbyEvent on SSE when there's a Pause or Stop stream event

ABeltramo committed:
fix: gracefully stop Gstreamer pipelines, updated zero copy config, upgraded gst-wayland-display

ABeltramo committed:
Committing clang-format changes

ABeltramo committed:
fix: better handling for Nvidia pipelines

This is still broken: when stopping a pipeline, OpenGL just trashes everything including our custom comp if it's running..

ABeltramo commented:
# Current status

Everything is pretty much wired together and working, see the short video below (youtube link):

[![Video showcasing co-op lobbies](http://img.youtube.com/vi/8an-SvnD4pk/0.jpg)](http://www.youtube.com/watch?v=8an-SvnD4pk)

There are two main issues remaining:
 - ~~On **AMD**: (I suppose Mesa in general) leaving a session on one resolution and joining from a device that asks for a bigger resolution causes a kernel panic. Not sure if this is specific to my dev machine, kernel version, etc.~~ **should be fixed now**
 - ~~On **Nvidia**: when any user stops a stream (either by leaving a lobby or by force closing from Moonlight) every other Wayland session running will crash.~~ fixed in https://github.com/games-on-whales/gst-wayland-display/pull/20

ABeltramo committed:
fix: automatic multi VAAPI GPUs, upgraded gst-wayland-comp

ABeltramo committed:
merge: merged latest changes from `stable`

ABeltramo committed:
Committing clang-format changes

ABeltramo committed:
fix: use introspection to automagically mount wolf socket

Marshy-Madness commented:
> # Current status
> 
> Everything is pretty much wired together and working, see the short video below (youtube link):
> 
> [![Video showcasing co-op lobbies](https://camo.githubusercontent.com/c8f6e8dba76ac8c02137c52e092c1df532e7359f05600ce70c154e5cd5454b07/687474703a2f2f696d672e796f75747562652e636f6d2f76692f38616e2d53766e4434706b2f302e6a7067)](http://www.youtube.com/watch?v=8an-SvnD4pk)
> 
> There are two main issues remaining:
> 
>     * ~On **AMD**: (I suppose Mesa in general) leaving a session on one resolution and joining from a device that asks for a bigger resolution causes a kernel panic. Not sure if this is specific to my dev machine, kernel version, etc.~ **should be fixed now**
> 
>     * On **Nvidia**: when any user stops a stream (either by leaving a lobby or by force closing from Moonlight) every other Wayland session running will crash.
>       
>       * This is tied to the fact that the zero copy video encoding pipeline needs to use some OpenGL elements (`glupload` and `glcolorconvert` to pass the DMA buffer into `nv*enc`) and when they get destroyed they'll trash some global context (I guess) that will mess up with our custom compositor (which uses OpenGL to render) causing it to crash.
>       * this issue is also present on `:stable`
>       * not sure how to get around this, I'll try to [share the OpenGL context](https://gstreamer.freedesktop.org/documentation/additional/design/opengl.html?gi-language=c) from `gst-wayland-comp` to the downstream `gl` elements, but I'm afraid it's [fairly complicated](https://discourse.gstreamer.org/t/shared-context-glmemory-and-glcolorconvert-filter-issue/1240/3)

Has this been fixed?

ABeltramo commented:
> Has this been fixed?

Nope, otherwise this would've been merged. :wink:

ABeltramo committed:
fix: remove hardcoded `WOLF_SOCKET_PATH` in startup.sh

ABeltramo committed:
fix: remove old Nvidia tweaks and fixed new Nvidia runtime

ABeltramo committed:
feat: new zero copy pipeline for Nvidia

brings in https://github.com/games-on-whales/gst-wayland-display/pull/20
and the forked gst-interpipe at https://github.com/games-on-whales/gst-interpipe

ABeltramo commented:
Just pushed a potential fix for the Nvidia issue, lots of changes.. Brings in github.com/games-on-whales/gst-wayland-display/pull/20 and the forked gst-interpipe at https://github.com/games-on-whales/gst-interpipe

TODO:
- [ ] we should be in full control over which Nvidia GPU gets used in a multi-GPU setup. Right now it's hardcoded to the device `0`, we should match the cuda device ID with the selected `render_node` instead.
- [ ] Test this properly! I gave it a light run with a couple sessions and lobbies, and it seems to work on my dev machine; let's see if it works when running in Docker too..

ABeltramo committed:
fix: upgrade gst-wayland-display to avoid not needed `gst-gl` dependency

Greaper88 commented:
For the multi-GPU setup, if say, you have 3 GPUs (my ultimate goal), would it be possible to auto-balance the load of the containers across the GPUs? IDK if that's possible to do dynamically, or if maybe it'd be possible to call the nvidia-smi command during launch to find the GPU with the lowest load and then assign that to the new container.

ABeltramo commented:
@Greaper88 yes, that is possible. We have full control over which GPU gets used for both the Wayland compositor and the video encoding pipeline so we could (in theory) automatically assign GPUs to different clients. It's definitely not implemented though..

ABeltramo committed:
fix: dynamically link CUDA, gracefully fallback when gstreamer-cuda-1.0 is not available

ABeltramo committed:
feat: upgrade Gstreamer to 1.26.7

ABeltramo committed:
feat: upgrade Boost to latest, suppress warnings when building libraries

ABeltramo committed:
feat: derive CUDA device ID from render node

ABeltramo committed:
Committing clang-format changes

ABeltramo committed:
fix: one single global Boost loader for cmake

ABeltramo committed:
fix: boost added missing process library

ABeltramo committed:
fix: joypad handling between lobbies and Moonlight sessions

ABeltramo committed:
feat: upgrade gst-wayland-display to latest

zexi committed:
fix: get CUDA device index in ascending order based on PCI addresses

ABeltramo committed:
fix: automatically detect Docker API version

ABeltramo committed:
fix: avoid exception when building gst-wayland-display without CUDA

ABeltramo committed:
fix: upgrade Rust and fix gst-wayland-display on CI

ABeltramo committed:
fix: upgrade Rust and fix gst-wayland-display on CI

ABeltramo committed:
Merge pull request #287 from zexi/get-cuda-id-by-render

fix: get CUDA device index in ascending order based on PCI addresses

ABeltramo commented:
Thanks everyone for following this amazing journey, it's time to merge this and start polishing and fixing the resulting issues. Time to start using `wolf:stable` everyone, see you on the other side!
